swarmlib_/aco4tsp/aco_algorithm.py

```diff
@@ -195,13 +204,39 @@ def _commit_paths(self, step, paths):
                            path, time_units=step - self._previous_commit)
         self._previous_commit = step
 
-    def _extract_uncommited_paths(self, paths: list[VechiclePath]):
+    def _extract_uncommited_paths(self, paths: list[VehiclePath]):
         return [full_path[len(vehicle.traveled_path)-1:]  # assumption is made, that commited and paths are in the same order
                 for vehicle, full_path in zip(self._vehicles, paths)
                 ]
 
-    def _calculate_distance(self, paths: list[VechiclePath]):
+    def _calculate_distance(self, paths: list[VehiclePath]):
         return sum(
             sum(self._env._routes_graph.get_edge_length((a, b))
                 for a, b in zip(p, p[1:])
                 ) for p in paths if len(p) > 1)
+    
+    def _calculate_threshold(self):
+        # average pheromone level
+        edges = self._env._routes_graph.get_edges()
+        return sum(self._env._routes_graph.get_edge_pheromone(edge) for edge in edges) / len(edges)
+
+    def _is_pheromone_level_significant(self, edge):
+        if self._env._routes_graph.get_edge_pheromone(edge) > self._calculate_threshold():
+          return True
+        return False
+
+    def _calculate_pheromone_ratio(self):
+        edges = list(filter(self._is_pheromone_level_significant, self._env._routes_graph.get_edges()))
+        return sum(self._env._routes_graph.get_edge_pheromone(edge) for edge in edges) / len(edges)
+
+    def _calculate_attractiveness_dispersion(self):
+        pheromone_values = list(map(self._env._routes_graph.get_edge_pheromone, self._env._routes_graph.get_edges()))
+        return np.std(pheromone_values)
+
+    def _calculate_attractiveness_ratio(self, best_paths: list[VehiclePath]):
+        pheromone_values = list(map(self._env._routes_graph.get_edge_pheromone, self._env._routes_graph.get_edges()))
+        best_path_total_pheromones = sum(
+            sum(self._env._routes_graph.get_edge_pheromone((a, b))
+                for a, b in zip(p, p[1:])
+                ) for p in best_paths if len(p) > 1)
+        return (best_path_total_pheromones / sum(pheromone_values))
```

Comment:
Nie miał to być stosunek feromonu na najlepszych ścieżkach do feromonu na pozostałych?

Czyli `best_path_total_pheromones / (sum(pheromone_values) - best_path_total_pheromones)`

![image](https://user-images.githubusercontent.com/21079319/236469957-c7c202cc-6cc2-45d2-a460-55555bcb2dfe.png)

Reply:
Tak, powinno być jeszcze to odejmowanie. Poprawiłam.